SongSnap_TextImageUnfrozen.ipynb

```diff
@@ -0,0 +1,4054 @@
+        "\n",
+        "        # Extract the last hidden states\n",
+        "        last_hidden_states = outputs.last_hidden_state\n",
+        "\n",
+        "        # Take the mean of the last hidden states (reduce it to 1 dimension)\n",
+        "        mean_last_hidden_state = torch.mean(last_hidden_states, dim=1)\n",
+        "\n",
+        "        # Pass the mean last hidden states through the fully connected layer (maps to size 8 embedding)\n",
+        "        output = self.fc(mean_last_hidden_state)\n",
+        "\n",
+        "        return output\n",
+        "\n",
+        "\n",
+        "\n",
+        "# Overall model that combines the embeddings from the three modalities and produces 4 outputs\n",
+        "# Separate fully connected layers for genre, popularity, danceability, and energy\n",
+        "class OverallModel(nn.Module):\n",
+        "    def __init__(self, num_modalities, num_genres):\n",
+        "        super(OverallModel, self).__init__()\n",
+        "        self.concatenated_output_size = 8 * num_modalities\n",
+        "        self.fc1 = nn.Linear(self.concatenated_output_size, 64)\n",
+        "        self.fc2_genre = nn.Linear(64, num_genres)\n",
+        "        self.fc3_popularity = nn.Linear(64, 1)\n",
+        "        self.fc4_danceability = nn.Linear(64, 1)\n",
+        "        self.fc5_energy = nn.Linear(64, 1)\n",
+        "\n",
+        "    def forward(self, sound_embedding, spec_embedding, text_embedding):\n",
+        "        # TODO: Concatenate with text and image embeddings\n",
+        "        # concatenated_embedding = torch.cat((sound_embedding, album_embedding, spec_embedding, text_embedding), dim=1)\n",
+        "        concatenated_embedding = torch.cat((sound_embedding, spec_embedding, text_embedding), dim=1)\n",
+        "\n",
+        "        x = torch.relu(self.fc1(concatenated_embedding))\n",
+        "        genre_output = torch.softmax(self.fc2_genre(x), dim=-1)\n",
+        "        popularity_output = torch.sigmoid(self.fc3_popularity(x))\n",
+        "        danceability_output = torch.sigmoid(self.fc4_danceability(x))\n",
+        "        energy_output = torch.sigmoid(self.fc5_energy(x))\n",
+        "        return genre_output, popularity_output, danceability_output, energy_output\n",
+        "\n",
+        "num_modalities = 3 # TODO: Change to 3 once we have text and image modalities\n",
+        "num_genres = num_unique_genres\n",
+        "\n",
+        "#TODO: Instantiate the remaining modality models\n",
+        "sound_model = SoundModalityModel(num_columns_X_sound).to(device)\n",
+        "# album_model = ImageModalityModel().to(device)\n",
+        "spec_model = ImageModalityModel().to(device)\n",
+        "text_model = TextModalityModel().to(device)\n",
+        "\n",
+        "overall_model = OverallModel(num_modalities, num_genres).to(device)\n",
+        "\n",
+        "\n",
+        "# TODO: Create optimizers for text and image modalities\n",
```

Comment:
wondering if there's any way we can improve the optimizers? this makes sense but maybe should use different LR, etc. for different optimizers since i think the models are different sizes